[AppService] functionapp: Added warning if customer tries to switch functionapp from premium to consumption.

gzuber opened the pull request:
Resolves #15694

**Description<!--Mandatory-->**  
<!--Why this PR? What is changed? What is the effect? etc. A high-quality description can accelerate the review process.-->

This PR involves changes to the `az functionapp update` command:

We're now requiring a user specify a `--force` flag to their command if they're attempting to migrate a functionapp from Premium -> Consumption. This is a potentially dangerous action if their functionapp uses Premium features that are not available in consumption. 

This also adds an error message if the user tries to migrate a functionapp to or from a Linux plan. Migration will only work on Windows at this time.

I also changed the exception handling slightly to add a helpful message if a migration attempt fails from the backend letting the customer know that some plans are incapable of upgrading.

**Testing Guide**  
<!--Example commands with explanations.-->

Most of these changes are warnings and error messages. The rest of these changes will either be tested by the move/migration tests we already have. 

**History Notes**  
<!--If your PR is not customer-facing, use {Component Name} in the PR title. Otherwise, use [Component Name] to allow our pipeline to add the title as a history note. If you need multiple history notes or would like to overwrite the note from the PR title, please fill in the following templates.-->

[AppService] BREAKING CHANGE: az functionapp update: Migrating a functionapp from Premium to Consumption plans now requires the '--force' flag.
[AppService] az functionapp update: Added error message if functionapp migration involves any plans on Linux.
[AppService] az functionapp update: Added more descriptive error message if functionapp migration fails.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

gzuber commented:
@cachai2 for help with wording the warning message. Part of me wonders if this is enough. It'll warn the user but too late for them to abort. I could see us adding warnings to the help menu or event making the user provide a `--force` flag or something like that. That would be a breaking change but considering how dangerous this is and how rarely it's probably used it might not be a big deal.

yonzhan commented:
AppService

cachai2 commented:
@gzuber thanks for working on this. I'll take a look and help out with wording as well to understand if we need to add more to ensure users have what they need. 

Also, is there any messaging for consumption to premium migration in cases where it might fail? We should provide some messaging in these cases as well to make sure customers aren't in the dark if their migration doesn't work.

cachai2 commented:
@gzuber I like most of the wording you had, and added a few updates. Please let me know your thoughts on the below for premium to consumption migration:

-----------
"Warning: Moving a functionapp from Premium to Consumption may result in loss of functionality and cause the app to break. 
Please ensure the functionapp is compatible with a Consumption plan and is not using any features only available in Premium.

"Do you still want to migrate the functionapp to a consumption plan?" yes/no
-----------

- I think it would be good to have a proactive check to allow users to decline the migration after seeing the error message but before the actual migration to give them the opportunity to verify. 
- I was also thinking about whether or not to include a link to docs. However, it could be a bit dangerous if anyone does the migration and something breaks that isn't in the docs, so it might be best to leave it out. Let me know your thoughts.

yungezz commented:
hi @qwordy could you pls help to review? thanks

qwordy commented:
/azp run

gzuber commented:
Thanks for reviewing! Just as an FYI I still have some changes to get in for this PR, hoping to get those pushed soon

gzuber commented on src/azure-cli/azure/cli/command_modules/appservice/commands.py:
```diff
@@ -32,28 +32,57 @@ def transform_web_list_output(webs):
+    except Exception:  # pylint: disable=broad-except
+        pass
+    return ex
 
```
My reasoning for refactoring this is because I wanted to take advantage of `_polish_bad_errors` because `az functionapp update` has been using this exception handler. However, I needed to add an additional message in case of an exception but not to all the other commands that use this handler. I believe this should preserve all the other commands' exception handling behavior.

gzuber commented:
@qwordy/@yungezz this should be ready for any additional review and merge (I added a few things), thanks!

gzuber commented:
@qwordy / @yungezz this should be ready for any additional review and merge (I added a few things), thanks!

qwordy commented:
@gzuber Error handling guidelines are updated. CLIError is not recommended to used any more. We suggest use more concrete error type. Thank you for your work.
https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md

gzuber commented:
@qwordy I changed the Errors for the only command I changed with this PR (`az functionapp update`). I didn't change `ex_handler_factory` because that would change the behavior of most of the commands in appservice. I think it's something we need to do, but I think that should be done in its own separate PR since it's such an extensive change.

Is there any chance we can merge this in time for [S178](https://github.com/Azure/azure-cli/milestone/96)? Please let me know if there are any more changes I need to make on my end, thanks!

qwordy commented:
> 
> 
> @qwordy I changed the Errors for the only command I changed with this PR (`az functionapp update`). I didn't change `ex_handler_factory` because that would change the behavior of most of the commands in appservice. I think it's something we need to do, but I think that should be done in its own separate PR since it's such an extensive change.
> 
> Is there any chance we can merge this in time for [S178](https://github.com/Azure/azure-cli/milestone/96)? Please let me know if there are any more changes I need to make on my end, thanks!

Thank you for your quick response. It is a tough job to update all existing code. We only require new code to raise more concrete error type now. I'll merge it.
